src/app/mauro/query-builder.service.ts

```diff
@@ -0,0 +1,145 @@
+import {
+  CatalogueItemDomainType,
+  DataType,
+  Profile,
+} from '@maurodatamapper/mdm-resources';
+import { QueryBuilderConfig } from 'angular2-query-builder';
+import { catchError, forkJoin, map, Observable, of } from 'rxjs';
+import {
+  DataElementSearchResult,
+  DataRequestQueryPayload,
+  QueryExpression,
+} from '../data-explorer/data-explorer.types';
+import { ProfileService } from './profile.service';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class QueryBuilderService {
+  constructor(private profileService: ProfileService) {}
+
+  private getQueryBuilderDatatype(dataType?: DataType): Observable<Profile> {
+    if (dataType?.domainType === CatalogueItemDomainType.PrimitiveType) {
+      const requestOptions = {
+        handleGetErrors: false,
+      };
+
+      return this.profileService.get(
+        CatalogueItemDomainType.PrimitiveType,
+        dataType?.id ?? '',
+        'uk.ac.ox.softeng.maurodatamapper.plugins.explorer.querybuilder',
+        'QueryBuilderPrimitiveTypeProfileProviderService',
+        requestOptions
+      );
+    }
+    return of({} as Profile);
+  }
+
+  public setupConfig(
+    dataElements: DataElementSearchResult[],
+    query?: DataRequestQueryPayload
+  ): Observable<
+    [
+      DataElementSearchResult[],
+      Required<DataRequestQueryPayload> | undefined,
+      QueryBuilderConfig
+    ]
+  > {
+    const requests$ = dataElements.map((dataElement) => {
+      const profile$ = this.getQueryBuilderDatatype(dataElement.dataType).pipe(
+        catchError((error) => {
+          if (error.status === 404) {
+            return of({} as Profile);
+          }
+
+          throw error;
+        })
+      );
+
+      const dataElement$ = of(dataElement);
+
+      return forkJoin([profile$, dataElement$]);
+    });
+
+    return forkJoin(requests$).pipe(
+      map((items) => {
+        const queryCondition = query
+          ? query.condition
+          : {
+              condition: 'and',
+              rules: [],
+            };
+
+        const config: QueryBuilderConfig = {
+          fields: {},
+        };
+
+        items.forEach(([data, dataElement]) => {
+          const dataTypeString =
+            dataElement.dataType?.enumerationValues?.length ?? 0 > 0
+              ? 'category'
+              : data?.sections?.length > 0
+              ? data.sections[0].fields[0].currentValue
+              : undefined;
+
+          if (dataTypeString) {
+            config.fields[dataElement.label] = {
+              name: dataElement.label + ' (' + dataTypeString + ')',
+              type: dataTypeString,
+              options: (dataElement.dataType?.enumerationValues ?? []).map(
+                (dataOption) => {
+                  return {
+                    name: dataOption.key,
+                    value: dataOption.value,
+                  };
+                }
+              ),
+              defaultValue:
+                dataTypeString.toLowerCase() === 'number'
```

Comment:
Similar case with this default value selection, this could potentially grow bigger.

Reply:
Private method added as suggested to maintain this.